render/render.js

```diff
@@ -415,37 +555,43 @@ module.exports = function($window) {
 	}
 
 	//remove
-	function removeNodes(vnodes, start, end, context) {
+	function removeNodes(vnodes, start, end, context, recycling) {
 		for (var i = start; i < end; i++) {
 			var vnode = vnodes[i]
 			if (vnode != null) {
 				if (vnode.skip) vnode.skip = false
-				else removeNode(vnode, context)
+				else removeNode(vnode, context, recycling)
 			}
 		}
 	}
-	function removeNode(vnode, context) {
+	// when a node is removed from a parent that's brought back from the pool, its hooks should
+	// not fire.
+	function removeNode(vnode, context, recycling) {
 		var expected = 1, called = 0
 		var original = vnode.state
-		if (vnode.attrs && typeof vnode.attrs.onbeforeremove === "function") {
-			var result = callHook.call(vnode.attrs.onbeforeremove, vnode)
-			if (result != null && typeof result.then === "function") {
-				expected++
-				result.then(continuation, continuation)
+		if (!recycling) {
```

Comment:
This would read cleaner (and have a much smaller diff) if you did `if (recycling) return` and dropped the nesting level back to where it was. Also, it could just as easily be done a couple lines above where it is.

Reply:
Looking more closely, we must still call the continuation, so we must define `expected` and `called`. I've moved the `original` assignment inside the conditional.